`lib/posix-process`: Implementation of `clone` system call

skuenzer opened the pull request:
This PR implements the `clone` system call which is used by [`lib/musl`](https://github.com/unikraft/lib-musl) for thread creation and which is also an important addition for supporting a Linux binary compatibility mode. The introduced architecture moves handling of particular clone flags to corresponding subsystem libraries so that our separation and configuration flexibility is kept.

This PR depends on PR #565.

skuenzer committed:
lib/nolibc: Place `errno` on the TLS

When scheduling is available, we place the `errno` variable on the
thread local storage.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Provide debug message option

Provides a configuration option to enable all debug messages within
the `posix-process` library.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Move deprecated implementation

Because we are going to replace step-wise the library implementation,
we move the current implementation to a file called `deprecated.c`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Initial POSIX process abstraction

This commits implements an initial POSIX process abstraction. Processes
have a process ID, can consist of one or multiple threads, and can have
a parent process. Threads are belonging to a process and get a thread
ID assigned.
The POSIX thread/process property can be enabled for a `uk_thread` by
calling `uk_posix_process_create()`. Children derive the activation from
the parent thread and get assigned to the same process as the parent.

Checkpatch-Ignore: INITIALISED_STATIC
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Remove `fork` and `vfork` stubs

Removes the unnecessary `ENOSYS` stubs of `fork` and `vfork`. `ENOSYS`
stubs are automatically created by the `syscall_shim` for unregistered
system calls.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/nolibc, lib/posix-process: Add `CLONE_*` flag definitions

Introduces the Linux-style clone flag definitions to `<sched.h>`. The
flags are derived from musl. Additional flag definitions are introduced
by `posix-process`. The library header defines flags that were not
defined by a given libC already.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: CLONE_PARENT_SETTID, CLONE_CHILD_SETTID, CLONE_THREAD

This commit introduces handler for `CLONE_PARENT_SETTID`,
`CLONE_CHILD_SETTID` ,`CLONE_THREAD`. `CLONE_PARENT_SETTID`,
`CLONE_CHILD_SETTID` store the child thread ID at a given location and
`CLONE_THREAD` is currently stubbed (but needed to be accepted in order
to support musl).

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: Handle CLONE_FILES flag

`CLONE_FILES` specifies if a cloned child shares the file descriptor
table with the parent. Because we support only a single and global table
at the moment, the handler requires `CLONE_FILES` to be set.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/vfscore: Handle CLONE_FS flag

`CLONE_FS` specifies if a cloned child shares the file system
information with the parent (mount points, chroot, etc.). Because, we
currently support only single file system information, the handler
requires `CLONE_FS` to be set.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uklock: Stub handler for CLONE_SYSVSEM flag

Introduces a stub handler for the `CLONE_SYSVSEM` flag. The flag is set
by musl during thread creation and is related to System V semaphores.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-futex: `SYS_set_tid_address`, CLONE_CHILD_CLEARTID flag

This commit provides an implementation for the `CLONE_CHILD_CLEARTID`
flag and introduces the related system call `SYS_set_tid_address`.
Whenever the child terminates, `0` is written to the given TID address
and a futex wake is issued to this address.

Checkpatch-Ignore: INITIALISED_STATIC
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Stub handler for CLONE_SIGHAND, CLONE_CLEAR_SIGHAND

The `CLONE_SIGHAND` flag and the `CLONE_CLEAR_SIGHAND` flag are
currently stubbed but required by `musl`. The stubbing handler checks
the flags for a valid combination according to the documentation
(man pages: clone(2)), like, `CLONE_SIGHAND` requiring `CLONE_VM` to be
set.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/nolibc: Linux-style errno numbers

Introduces Linux-style errno numbers (`errno.h`) and error message
strings (`string.c`). This is done so that we achieve binary
compatibility with Linux errno codes that may be returned by system
calls. Both lists are generated with the `errno` command line tool
(https://manpages.debian.org/testing/moreutils/errno.1.en.html)
with the following bash command:

* `errno.h`:
```bash
errno -ls | \
awk '{ printf "#define "$1"\t\t"$2; ' \
  'printf " /* "substr($0, length($1) + length($2) + 3, 1024)" */\n"; }'
```

* `string.c`
```bash
for (( I=0; I<256; I++ )); do
  errno $I | awk '{ printf "\tcase "$1":\n"; '
             'printf "\t\tstrerr = \""substr($0, '
             'length($1) + length($2) + 3, 1024)"\";\n\t\tbreak;\n"; }';
done
```

Checkpatch-Ignore: ENOSYS
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Ignore CLONE_DETACHED flag

This commits provides a handler for the historical `CLONE_DETACHED` flag
that existed during Linux 2.5 development and is ignored since then.
From man pages (clone(2)):
 "CLONE_DETACHED (historical)
   For a while (during the Linux 2.5 development series)
   there was a CLONE_DETACHED flag, which caused the parent
   not to receive a signal when the child terminated.
   Ultimately, the effect of this flag was subsumed under the
   CLONE_THREAD flag and by the time Linux 2.6.0 was
   released, this flag had no effect.  Starting in Linux
   2.6.2, the need to give this flag together with
   CLONE_THREAD disappeared.
   This flag is still defined, but it is usually ignored when
   calling clone(). However, see the description of
   CLONE_PIDFD for some exceptions."

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Stubs for wait system calls

This commit introduces stubs for the wait system calls. It is a skeleton
in preparation of a rewrite.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

dragosargint reviewed (commented):
Everything looks good from my point of view. I'm waiting for the `DRAFT` tag to be removed before I add the reviewed-by tag. Here are the tests that I used for this PR:
https://github.com/dragosargint/test

skuenzer committed:
lib/posix-process: Initial clone system call implementation

This commits implements the low-level version of the clone and clone3
system calls.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Skeleton for `clone` libc wrapper

This commit introduces a skeleton for `clone` libc wrapper.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Clone flag `init` and `term` hooks

Similar to the `ctortab` or `inittab`, this commit introduces an inittab
that is used to register handlers that can get involved during a clone
system call depending on given clone flags. This is done for achieving a
proper feature separation with our libraries. For instance,
`CLONE_FILES` specifies if a child shares the file descriptor table
with the parent. Such functionality should be implemented in `vfscore`
(or `posix-fdtab` in the future) instead of `posix-process`.

Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Debug message of clone flags and parameters

This commit introduces debug messages that print the called parameters
and flags of a clone system call.

Checkpatch-Ignore: TRAILING_STATEMENTS
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/posix-process: Handle CLONE_VM flag

At the moment, Unikraft does not support multiple process address
spaces. Because of this, the `CLONE_VM` flag is checked for its
presence. If it is missing, the clone system call will return an error.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

craciunoiuc commented on lib/posix-process/Makefile.uk:
```diff
@@ -6,23 +6,30 @@ LIBPOSIX_PROCESS_COMMON_INCLUDES-y     += -I$(LIBPOSIX_PROCESS_BASE)/musl-import
+LIBPOSIX_PROCESS_SRCS-y += $(LIBPOSIX_PROCESS_BASE)/signals.c
+LIBPOSIX_PROCESS_SRCS-$(CONFIG_LIBPOSIX_PROCESS_CLONE) += $(LIBPOSIX_PROCESS_BASE)/clone.c
+LIBPOSIX_PROCESS_SRCS-$(CONFIG_LIBPOSIX_PROCESS_CLONE) += $(LIBPOSIX_PROCESS_BASE)/clonetab.ld
+COMPFLAGS-$(CONFIG_LIBPOSIX_PROCESS_PIDS) += -fno-builtin-exit -fno-builtin-exit-group
```
```suggestion
LIBPOSIX_PROCESS_COMPFLAGS-$(CONFIG_LIBPOSIX_PROCESS_PIDS) += -fno-builtin-exit -fno-builtin-exit-group
```

This doesn't really follow the naming scheme

skuenzer replied:
These flags are important to set globally (list of compiler flags, language independent: `COMPFLAGS-y`) because otherwise GCC is inlining a built-in for exit instead of our function.

craciunoiuc commented on lib/posix-process/clone.c:
```diff
@@ -0,0 +1,569 @@
+		 * user land), we allocate an Unikraft TLS because Unikraft
+		 * places TLS variables and uses them effectively as TCB.
+		 */
+#if CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS
+		if (uk_syscall_ultlsp() != 0x0) {
+			uk_pr_debug("Allocating an Unikraft TLS for the new child, parent called from context with custom TLS\n");
+		} else
+#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
+		{
+			uk_pr_debug("Allocating an Unikraft TLS for the new child, no TLS given by parent\n");
+		}
```
```suggestion
#if CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS
		if (uk_syscall_ultlsp() != 0x0) {
			uk_pr_debug("Allocating an Unikraft TLS for the new child, parent called from context with custom TLS\n");
		} else {
			uk_pr_debug("Allocating an Unikraft TLS for the new child, no TLS given by parent\n");
		}
#endif /* CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS */
```

craciunoiuc replied:
Is this okay? I don't get the placement there

skuenzer replied:
My intention here is to compile-in the `if-else` with this configuration option set. If the option is not set, the message of the else block should stay (the other case is not supported without that option).

craciunoiuc commented on lib/posix-process/clone.c:
```diff
@@ -0,0 +1,569 @@
+	/* We will return the child's thread ID in the parent */
+	ret = ukthread2tid(child);
+
+	/* Assing the child to a scheduler and give it the chance
```
```suggestion
	/* Assessing the child to a scheduler and give it the chance
```

skuenzer replied:
:+1: yes, I am fixing this typo: `Assing` -> `Assign`

craciunoiuc reviewed (changes requested):
`lib/posix-process: process.c: Handle CLONE_PARENT_SETTID, CLONE_CHILD_SETTID, CLONE_THREAD` -> too long
` lib/posix-process: Handle CLONE_SIGHAND, CLONE_CLEAR_SIGHAND (be prepared that uksignal will get moved into posix-process)` -> too long
`lib/uklock: Add stub for system V sempahores: CLONE_SYSVSEM` -> `lib/uklock: Add stub for system V semaphores: CLONE_SYSVSEM`


Hey @skuenzer,

An initial set of comments from my side. I'll do a final review once this is ready.

skuenzer commented:
> `lib/posix-process: process.c: Handle CLONE_PARENT_SETTID, CLONE_CHILD_SETTID, CLONE_THREAD` -> too long ` lib/posix-process: Handle CLONE_SIGHAND, CLONE_CLEAR_SIGHAND (be prepared that uksignal will get moved into posix-process)` -> too long `lib/uklock: Add stub for system V sempahores: CLONE_SYSVSEM` -> `lib/uklock: Add stub for system V semaphores: CLONE_SYSVSEM`
> 
> Hey @skuenzer,
> 
> An initial set of comments from my side. I'll do a final review once this is ready.

Hey @craciunoiuc, thanks a lot for the first check. I pushed now the first non-draft version of the PR. I hope there are not too many typos left - this reminds me to properly setup the spellchecker on this machine. 😉 

Thanks, Simon

razvand commented:
@skuenzer , there is an issue with `wait()` and `waitpid()` in `lib/posix-process/deprecated.c`:

```
/usr/bin/ld: /home/unikraft/razvand/scripts/workdir/apps/app-redis/build/libmusl.o: in function `waitpid':
/home/unikraft/razvand/scripts/workdir/apps/app-redis/build/libmusl/origin/musl-1.1.19//src/process/waitpid.c:6: multiple definition of `waitpid'; /home/unikraft/razvand/scripts/workd
ir/apps/app-redis/build/libposix_process.o:/home/unikraft/razvand/scripts/workdir/unikraft/lib/posix-process/deprecated.c:188: first defined here
/usr/bin/ld: /home/unikraft/razvand/scripts/workdir/apps/app-redis/build/libmusl.o: in function `wait':
/home/unikraft/razvand/scripts/workdir/apps/app-redis/build/libmusl/origin/musl-1.1.19//src/process/wait.c:4: multiple definition of `wait'; /home/unikraft/razvand/scripts/workdir/app
s/app-redis/build/libposix_process.o:/home/unikraft/razvand/scripts/workdir/unikraft/lib/posix-process/deprecated.c:181: first defined here
collect2: error: ld returned 1 exit status
```

I created a solution [here](https://github.com/unikraft-upb/unikraft/commit/4508cfdd1166cdceac82c148474cb3383fbccd00). I think it's best if you updated this PR to use this update.

craciunoiuc commented on lib/posix-process/process.c:
```diff
@@ -30,435 +33,622 @@
+		return -EINVAL;
 
-	return rc;
+	*((pid_t *) cl_args->parent_tid) = child_tid;
```
Why is the child_tid stored in the parent_tid?

skuenzer replied:
I also dislike the naming: `parent_tid` (from clone arguments) is a given memory address in the parent where the child's thread ID should stored. https://man7.org/linux/man-pages/man2/clone.2.html: `CLONE_PARENT_SETTID`

craciunoiuc commented on lib/posix-process/signals.c:
```diff
@@ -0,0 +1,64 @@
+		     == (CLONE_SIGHAND | CLONE_CLEAR_SIGHAND)))
+		return -EINVAL;
+	/* CLONE_SIGHAND requires CLONE_VM */
+	if (unlikely(!(!(cl_args->flags & CLONE_SIGHAND))
+		     && !(cl_args->flags & CLONE_VM)))
```
I will use 100% of my brain power to rewrite this:
https://en.wikipedia.org/wiki/De_Morgan%27s_laws

Double check that it's okay and do this below also (and in other places that it appears. We basically save 3 operations) 😅 

```suggestion
	if (unlikely((cl_args->flags & CLONE_SIGHAND)
		     || (cl_args->flags & CLONE_VM)))
```

skuenzer replied:
😉 Looks okay to me. This optimization would have been done by the compiler for sure but it increases readability. 👍🏻

skuenzer replied:
Ah wait, this is different. The first argument is double negated which normalizes its binary representation. This is not necessary here. I was looking at the `||` expression and wondered that the whol eboolean expression doe snot fit anymore.

craciunoiuc commented on lib/posix-process/signals.c:
```diff
@@ -0,0 +1,64 @@
+		return -EINVAL;
+	/* CLONE_THREAD requires CLONE_SIGHAND */
+	if (unlikely(!(!(cl_args->flags & CLONE_THREAD))
+		     && !(cl_args->flags & CLONE_SIGHAND)))
```
also here

craciunoiuc replied:
there's multiple, places, I won't mark them anymore

skuenzer replied:
same here - double negation ;-) I am simplifying it.

craciunoiuc reviewed (commented):
A final round of comments from my side

dragosargint reviewed (approved):
I retested this PR and it works fine! Thanks, @skuenzer!  I have no other comments to add, so I will approve it
Reviewed-by: Dragos Iulian Argint <dragosargint21@gmail.com>

craciunoiuc reviewed (approved):
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

adinasm reviewed (approved):
Looks good to me too.

Reviewed-by: Adina Smeu <adina.smeu@gmail.com>
